feat(form-workspaces-be/5): add delete workspace logic

hanstirtaputra opened the pull request:
Builds on #4252 

Part of #4039 

## Solution
Add logic for DELETE `api/v3/admin/workspaces/{workspaceId}` endpoint

**Breaking Changes** 
- [x] No - this PR is backwards compatible

hanstirtaputra committed:
feat: add delete workspace functionality

hanstirtaputra committed:
feat: add delete workspace functionality tests

mantariksh commented on src/app/modules/workspace/workspace.service.ts:
```diff
@@ -80,12 +81,28 @@ export const updateWorkspaceTitle = (
 
 export const deleteWorkspace = (
```
would be good to use an object arg here as well

mantariksh commented on src/app/models/workspace.server.model.ts:
```diff
@@ -70,6 +75,34 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
+    })
+    const session = await this.startSession()
+
+    session.startTransaction()
+    const deleted = await this.deleteOne({
+      _id: workspaceId,
+      admin: admin,
+    })
+    if (shouldDeleteForms) {
+      await Form.updateMany(
+        { _id: { $in: workspaceToDelete?.formIds } },
+        { $set: { status: FormStatus.Archived } },
+      )
+    }
+    await session.commitTransaction()
+    session.endSession()
```
this logic belongs in the service layer above, not in the model layer.

model methods are meant to deal with operations on that model itself, without knowledge of the business logic. services are meant to coordinate changes to the models according to the business logic.

if you check the other usage of transactions in the codebase (`createTwilioCredentials`) you can see that it takes place in the service layer as well.

hanstirtaputra replied:
Updated to move the transaction logic to the service layer!

hanstirtaputra committed:
refactor: use object function argument for deleteWorkspace

hanstirtaputra committed:
fix: move deleteWorkspace transaction logic to WorkspaceService

hanstirtaputra committed:
feat: update deleteWorkspace to return boolean instead of void

hanstirtaputra committed:
refactor: deleteWorkspace transaction to use mongoose withTransaction

hanstirtaputra committed:
fix: use form document method to archive forms instead of static method

This resolves the issue of failing transaction due to 'nearest' readPreference for FormSchema.
Setting readPreference for static queries throws an error.

hanstirtaputra committed:
fix: tests for deleteWorkspace functionality

hanstirtaputra committed:
refactor: remove admin field from deleteWorkspace

mantariksh commented on src/types/workspace.ts:
```diff
@@ -29,4 +29,12 @@ export interface IWorkspaceModel extends Model<IWorkspaceSchema> {
+    session,
+  }: {
+    workspaceId: IWorkspaceSchema['_id']
+    session?: ClientSession
```
if we expect this operation to always take place within a transaction, we should make this mandatory

hanstirtaputra replied:
Currently the test environment database doesn't support transactions, so I made the session optional so that we can mock the implementation of this function in the test environment (to run without transaction).

I've added a comment to this to make it clearer!

Can see issue #4503 for details

mantariksh replied:
hmm, we shouldn't compromise on type safety in production code. for the dev taking over this - is there a better way to do this, e.g. by creating a mock session object to use for tests?

foochifa replied:
Agreed on this - and we might be able to use our current `mongodb-memory-server-core` [package](https://github.com/nodkz/mongodb-memory-server#replica-set-start) to create a replica set on our DB. Which may help us test for transaction.

However, this branch is too deprecated, will merge it into `form-v2/form-workspaces` first and work on this issue later. 

FRM-976

mantariksh commented on src/app/modules/workspace/workspace.service.ts:
```diff
@@ -83,14 +84,65 @@ export const updateWorkspaceTitle = ({
+}): Promise<void> => {
+  const workspaceToDelete = await WorkspaceModel.findOne({
+    _id: workspaceId,
+  })
```
why do we retrieve the whole document even when `shouldDeleteForms` is false?

hanstirtaputra replied:
Updated this to use `findOneAndRemove` instead so we get the whole document when we delete a workspace

mantariksh commented on src/app/modules/form/admin-form/admin-form.service.ts:
```diff
@@ -1498,3 +1498,17 @@ const deleteTwilioTransaction = async (
+    _id: { $in: formIds },
+  })
+
+  canBeArchivedForms.forEach(async (form) => await form.archive(session))
```
rather than archive each form in an independent async call, we should issue one database query to archive multiple form IDs. that's what you did when profiling the time taken for this query, correct?

hanstirtaputra replied:
> that's what you did when profiling the time taken for this query, correct?

Yes that's what I did.

I updated the implementation to use one db query!

mantariksh commented on src/app/models/__tests__/workspace.server.model.spec.ts:
```diff
@@ -199,5 +203,37 @@ describe('Workspace Model', () => {
     })
+
+    describe('deleteWorkspace', () => {
+      it('workspace should no loger be in database when deleted', async () => {
```
test descriptions normally use the "should...when" pattern, e.g. "should correctly remove workspace from database when deleted"

mantariksh commented on src/app/models/workspace.server.model.ts:
```diff
@@ -70,6 +70,22 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
+      },
+      { session },
+    )
+    return deleted.deletedCount == 1
```
```suggestion
    return deleted.deletedCount === 1
```

mantariksh commented on src/app/models/workspace.server.model.ts:
```diff
@@ -70,6 +70,22 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
+    workspaceId: IWorkspaceSchema['_id']
+    session?: ClientSession
+  }) {
+    const deleted = await this.deleteOne(
```
why not use `findOneAndRemove` so we can return the deleted document, which is more informative for the caller?

mantariksh commented on src/app/modules/workspace/__tests__/workspace.service.spec.ts:
```diff
@@ -275,4 +278,181 @@ describe('workspace.service', () => {
+      count: 0,
+    }
+
+    it('should return undefined when successfully deleted workspace', async () => {
```
usually we should think carefully about returning `undefined`, as it's an uninformative response type. one alternative could be to use `findOneAndRemove` in the db layer so as to be able to at least return the deleted document.

hanstirtaputra replied:
Updated to use `findOneAndRemove`!

mantariksh commented on src/app/modules/workspace/workspace.service.ts:
```diff
@@ -83,14 +84,65 @@ export const updateWorkspaceTitle = ({
+}: {
+  workspaceId: string
+  shouldDeleteForms: boolean
+}): ResultAsync<void, DatabaseError> => {
```
this should be `PossibleDatabaseError` as per the return type of `transformMongoError` right?

hanstirtaputra replied:
Yes, updated the type to `PossibleDatabaseError`

mantariksh commented on src/app/modules/workspace/__tests__/workspace.service.spec.ts:
```diff
@@ -275,4 +278,181 @@ describe('workspace.service', () => {
+      expect(actual._unsafeUnwrap()).toEqual(undefined)
+    })
+
+    it('should return undefined when failed to delete workspace', async () => {
```
this illustrates the previous comment - we return the same thing regardless of the outcome of the deletion

mantariksh commented on src/app/modules/workspace/__tests__/workspace.service.spec.ts:
```diff
@@ -275,4 +278,181 @@ describe('workspace.service', () => {
+        workspaceId: mockWorkspace._id.toHexString(),
+        shouldDeleteForms: false,
+      })
+      const doesFormExist = await FormModel.exists({ _id: mockFormId })
+      const isFormArchived = await FormModel.exists({
+        _id: mockFormId,
+        status: FormStatus.Archived,
+      })
```
instead of making two async calls, we can just retrieve the form once and check that it is not null and that status is not archived

hanstirtaputra replied:
Updated to do this instead! Thanks

mantariksh commented on src/app/modules/workspace/__tests__/workspace.service.spec.ts:
```diff
@@ -275,4 +278,181 @@ describe('workspace.service', () => {
+      expect(isFormArchived).toEqual(false)
+    })
+
+    it('should not archive forms or delete workspace when transaction fails at forms deletion', async () => {
```
well done on covering all these cases!

mantariksh commented:
how do we ensure that when the user opts to delete the forms within a workspace, only the forms that they own are deleted?

hanstirtaputra committed:
fix: delete workspace transaction to update model instead of documents

Fixed the readPreference issue in multi-document transactions by setting
readPreference in updateMany function on FormModel to be 'primary'

hanstirtaputra committed:
refactor: workspace service test to reduce db queries

hanstirtaputra committed:
fix: remove unused session parameter from form model archive method

hanstirtaputra commented:
> how do we ensure that when the user opts to delete the forms within a workspace, only the forms that they own are deleted?

Added a check in AdminFormService's `archiveForms` function to only archive forms the user is an admin of

mantariksh commented:
closing this to clean up the open PRs until we revisit it!

foochifa commented:
merging this to feature branch to rebase to current stack. Thanks @hanstirtaputra! Will also follow up on #4448
